main.py

```diff
@@ -183,92 +194,122 @@ def parse_scaffolding_result(output):
 def initiate_code_modification(code_files, modification_ask):
   new_code_files = []
+
+  # iterate over the code files and generate a new code block for each file using the modification prompt
   for file_code_pair in code_files:
     mod_prompt = get_modification_prompt("\n".join(file_code_pair["code_block"].split("\n")[1:]), modification_ask)
     modification_result = generate(mod_prompt)
     print("MOD_RESULT:", modification_result)
+
+    # if the modification result contains a new code block, update the code block for the current file
     if "RETURNEDCODE" in modification_result:
       modification_result = modification_result[modification_result.index("[/INST]"):]
       code_block_raw_string = modification_result[modification_result.index("RETURNEDCODE") + len("RETURNEDCODE"):]
       file_code_pair["code_block"] = re.findall(r"```(.*?)```", code_block_raw_string, re.DOTALL)[0]
     new_code_files.append(file_code_pair)
-  return new_code_files
 
+  return new_code_files
 
+# this function takes a list of code files and generates a new package.json file if any dependencies are missing
+# returns the new package.json file as a string
 def resolve_missing_dependencies(code_files):
+  print("Resolving missing dependencies...")
+  # combine all code files into a single string
   print("Resolving missing dependencies...")
   codebase = "\n".join(
     list(map(lambda x: f"FILENAME: {x['file_name']}\nSOURCE: {x['code_block']}\n", code_files))
   )
+  # generate a prompt to resolve missing dependencies
   dep_prompt = get_dependency_prompt(codebase)
+  # use Llama 2 model to generate a response to the prompt
   dep_result = generate(dep_prompt)
+  # extract the package.json file from the response
   dep_result = dep_result[dep_result.index("[/INST]"):]
   print(dep_result)
+
+  # if a new package.json file was generated, return it
   if "PACKAGEJSON" in dep_result:
     package_json_text = re.findall(r"```(.*?)```", dep_result, re.DOTALL)[0]
     return package_json_text
   else:
     return None
 
-def dev_loop(code_files, user_ask, modification_ask=None):
+# this function takes a list of code files and a modification prompt and generates a new code block for each file using the prompt
+def dev_loop(code_files, user_ask, modification_ask=None, modify_all=False):
   if modification_ask:
-    # update each related code block with a prediction using the modification ask of the user
-    code_files = initiate_code_modification(code_files, modification_ask)
-
-  # dependency resolving
+    if modify_all:
+      # update all code blocks with a prediction using the modification ask of the user
+      code_files = initiate_code_modification(code_files, modification_ask)
+    else:
+      # update each related code block with a prediction using the modification ask of the user
+      for code_file in code_files:
+        if modification_ask in code_file["code_block"]:
```

Comment:
So far the comments are solid for cleaning it up. But not fully sure what this part is doing? This is checking if the modification ask is a substring in the code? Might be better to have some way to detect whether the code matches the modification ask. Could maybe vectorize the code and find matches.

Reply:
Sorry, just woke up myself (on UK time so am a bit delayed in responding). 

So yeah, it checks if it is a substring in the code. I'm not a fan of this way of doing it, as nested loops like this make it look clunky and feel unnecessary. Vectorizing the code to find matches is probably the better way of doing it. I've just put in a push to my branch that includes a rudimentary solution to vectorizing & finding if the code block matches the modification ask using 2 new functions:

- `vectorize_code` <-- This should vectorize the code block, and return a vector representation of the code block itself
- `find_closest_matches`  <-- This should find the closest matches to the modification prompt, and then return the indices of the closest matches to be used in the reworked `dev_loop` function

I'm only a junior, so still learning and this implementation might not be the best way of doing it!

Reply:
I do like the creativity in your vectorization implementation! Good job.

However, one thing is we want to use sentence transformers models for the vectorization here. And make it so people can get embeddings with any transformers model they'd like. Honestly I have to think through that feature more. Do you mind removing that part with the modification ask work/vectorization and I'll approve. I like the optional summarization and the clean up part you did

Reply:
Removed it, and pushed. It shold be ready for merge now.